Add week 4 learning outcomes (#8)
foundersandcoders/coursebook: learning-outcomes.md

# Week 4 Learning Outcomes

## Node

### Servers

- [ ] **We can create a web server that responds to http requests**
- [ ] We can conditionally set ports for our server based on the runtime environment
- [ ] We can read secrets from a .env file
- [ ] We can prevent sensitive data from being pushed to GitHub

### GET requests

- [ ] We can route requests to the correct handler function(s)
- [ ] We can serve different types of files to the client
- [ ] We can read information sent in a URL query string

> ### POST requests

Reviewer: The `GET` section is stuff that applies to any HTTP method really. Most of the `POST` stuff is about handling HTTP bodies, which lots of methods can have.

Could we instead put the `GET` stuff under "Servers" and have the specific "submitting data" LOs under "Request bodies" or something?

- [ ] We can handle data received from a POST request

> - [ ] We can ‘chunk’ data sent from the client side on the server side

Reviewer: This might be misleading—the request body arrives in chunks that we have to handle. We don't actually chunk the data ourselves

```suggestion
- [ ] We can handle streams of data sent from the client that don't arrive all at once
```

> - [ ] We can give descriptive names to HTML form input fields so that it is easy to access data in the POST request object

Reviewer: ```suggestion
- [ ] We can give descriptive names to HTML form input fields so that it is easy to access data in the request
```
Forms make `GET` requests by default (see [my new workshop](https://github.com/oliverjam/learn-html-forms)) so this should probably be a more general LO

- [ ] We can parse a request object

> - [ ] We can stringify a response object

Reviewer: What exactly does this mean? Responding to the browser with an HTML/JSON string?

Req/res stuff isn't really specific to POST requests, so maybe we don't want them in this section?

### Core modules

- [ ] **We can use core Node modules**
- [ ] We can require core modules in our files

> - [ ] We can access and invoke methods provided by core modules

Reviewer: These three kind of say the same thing

- [ ] We can read files synchronously and asynchronously using Node’s fs module

> ## Design LOs

Reviewer: ```suggestion
## Design
```

- [ ] **We can design a website that is easy to use and to navigate**
- [ ] We can identify common design patterns that aid usability
- [ ] We can build website navigation that demonstrates good usability
- [ ] We can make HTML elements look different on hover and focus to indicate their state to the user

> ## Testing LOs

Reviewer: ```suggestion
## Testing
```
